pkg/pool-weighted/contracts/WeightedPool.sol

```diff
@@ -401,4 +403,51 @@ contract WeightedPool is BaseWeightedPool, WeightedPoolProtocolFees {
     {
         return super._isOwnerOnlyAction(actionId);
     }
+
+    function getSwapFeePercentage(bytes memory userData,
+     OperationType _operation) public view virtual override returns (uint256 _fee) {
+        // using tx.origin insted of msg.sender as these functions are called 
+        // during join/exit/swap via vault
+        if(isCustomFeeEnabled && canSetCustomFee(tx.origin)){
+            if(_operation == OperationType.JOIN ){
+                WeightedPoolUserData.JoinKind kind = userData.joinKind();
+                if(kind == WeightedPoolUserData.JoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT){
+                    _fee = userData.exactTokensInForBptOutCustomFee();
+                } else {
+                if(kind == WeightedPoolUserData.JoinKind.TOKEN_IN_FOR_EXACT_BPT_OUT){
+                    _fee =  userData.tokenInForExactBptOutCustomFee();
+                }
+                }
+            }else{
+                if(_operation == OperationType.EXIT){
+                    WeightedPoolUserData.ExitKind kind = userData.exitKind();
+                    if(kind == WeightedPoolUserData.ExitKind.EXACT_BPT_IN_FOR_ONE_TOKEN_OUT){
+                        _fee =  userData.exactBptInForTokenOutCustomFee();
+                    } else {
+                    if(kind == WeightedPoolUserData.ExitKind.BPT_IN_FOR_EXACT_TOKENS_OUT){
+                        _fee =  userData.bptInForExactTokensOutCustomFee();
+                    }
+                    }
+                } else{
+                    if(_operation == OperationType.SWAP){
+                        _fee = userData.swapCustomFee();
+                    }
+                }
+            }
+        }else{
+            _fee = getSwapFeePercentage();
+        }
+    }
+
+    function setSwapFeePercentage(uint256 swapFeePercentage) public virtual override whenNotPaused {
+        // here msg.sender is used as this function directlly be called by the admin
+        // not via any other contract
+        if (isCustomFeeEnabled) {
+            require(solver == msg.sender,'CALLER_IS_NOT_SOLVER');
+             _setSwapFeePercentage(swapFeePercentage);
+        } else {
+            super.setSwapFeePercentage(swapFeePercentage);
```

Comment:
If `isCustomFeeEnabled` is going to be immutable / true, that would break the capability of Governance of setting fees to any pools created from here. 

Should Governance always be capable of setting the fee anyways? I'd consider using something like `authenticateFor`.

Reply:
No, governance should not be able to set the fess. The solver has certain expectations regarding this pool's fee, governance being able to set the fee could potentially rug the solver